electrumsv/network_support/esv_client.py

```diff
@@ -0,0 +1,436 @@
+
+    # ----- General Websocket ----- #
+    async def _fetch_peer_channel_message_job(self,
+            peer_channel_notification_queue: asyncio.Queue[GeneralNotification]) -> None:
+        """Can run multiple of these concurrently to fetch new peer channel messages"""
+        message: GeneralNotification = await peer_channel_notification_queue.get()
+        channel_id: ChannelId = message['result']['id']
+
+        peer_channel = await self.get_single_peer_channel_cached(channel_id)
+        messages: Optional[list[PeerChannelMessage]] = await peer_channel.get_messages()
+        if messages is not None:
+            for pc_message in messages:
+                if pc_message['content_type'] == 'application/json':
+                    # Todo should probably check for PeerChannelType.MERCHANT_API before type cast
+                    json_payload: MAPICallbackResponse = cast(MAPICallbackResponse,
+                        pc_message['payload'])
+                    if json_payload.get("callbackReason") \
+                            and json_payload["callbackReason"] == "merkleProof"\
+                            or json_payload["callbackReason"] == "doubleSpendAttempt":
+                        self._merkle_proofs_queue.put_nowait(json_payload)
+                    else:
+                        logger.error(f"PeerChannelMessage not recognised: {pc_message}")
+
+                if pc_message['content_type'] == 'application/octet-stream':
+                    logger.error(f"Binary format PeerChannelMessage received - not supported yet")
+        else:
+            logger.error("No messages could be returned from channel_id: %s, "
+                         "do you have a valid read token?", channel_id)
+
+    async def _message_fetcher_job(self) -> None:
+        """Idempotent - if spawned twice, the second time will do nothing"""
+        if not self._message_fetcher_is_alive:
+            peer_channel_notification_queue: asyncio.Queue[GeneralNotification] = asyncio.Queue()
+            for i in range(self._FETCH_JOBS_COUNT):
+                asyncio.create_task(
+                    self._fetch_peer_channel_message_job(peer_channel_notification_queue))
+
+            async for notification in self.subscribe_to_general_notifications():
+                peer_channel_notification_queue.put_nowait(notification)
+
+    async def wait_for_merkle_proofs_and_double_spends(self) -> AsyncIterable[TSCMerkleProof]:
+        if not self._message_fetcher_is_alive:
+            asyncio.create_task(self._message_fetcher_job())
+
+        # https://github.com/bitcoin-sv-specs/brfc-merchantapi#callback-notifications
+        while True:
+            # Todo run select query on MAPIBroadcastCallbacks to get libsodium encryption key
+            callback_response = await self._merkle_proofs_queue.get()
+            tsc_merkle_proof: TSCMerkleProof = json.loads(callback_response['callbackPayload'])
```

Comment:
`TSCMerkleProof` is a dataclass. We need to parse the proof from json and keep it in the `TSCMerkleProof` binary oriented format.

Reply:
This was due to a name clash with my own TypeDict class and it escaped my notice. Will fix.